src/Interpreters/Set.cpp

```diff
@@ -473,7 +475,7 @@ MergeTreeSetIndex::MergeTreeSetIndex(const Columns & set_elements, std::vector<K
     ::sort(indexes_mapping.begin(), indexes_mapping.end(),
         [](const KeyTuplePositionMapping & l, const KeyTuplePositionMapping & r)
         {
-            return std::tie(l.key_index, l.tuple_index) < std::tie(r.key_index, r.tuple_index);
+            return std::forward_as_tuple(l.key_index, l.rhs_chain.size(), l.lhs_chain.size(), l.tuple_index) < std::forward_as_tuple(r.key_index, r.rhs_chain.size(), r.lhs_chain.size(), r.tuple_index);
```

Comment:
This is a bit tricky line. Looks like we prefer the shortest monotonic chain?
Let's add a comment about it (and why).

Reply:
As I understand, in this logic, if a key column appear in multiple positions in the lhs of IN, then we will only chose one of them. If we have `a` and `f(a)`, then doesn't it will be better to choose the condition on `a` instead of `f(a)`? Pls correct me if I'm wrong.

Reply:
I added a comment to explain